File: 2375._Construct_Smallest_Number_From_DI_String.cpp
// https: // leetcode.com/problems/construct-smallest-number-from-di-string/description/?envType=daily-question&envId=2025-02-18
// Solution taken from: https://www.youtube.com/watch?v=45n_7EC1N4s&ab_channel=Techdose
// Did I understand properly? No

class Solution
{
public:
    string smallestNumber(string pattern)
    {
        int n = pattern.size();
        stack<int> st;
        string res = "";

        for (int i = 1; i <= n + 1; i++)
        {
            st.push(i);
            char c = pattern[i - 1];

            // If the pattern is increasing then reverse it and save
            if (i == n + 1 || c == 'I')
            {
                while (!st.empty())
                {
                    res.push_back(char('0' + st.top()));
                    st.pop();
                }
            }
        }
        return res;
    }
};

File: 2375._Construct_Smallest_Number_From_DI_String.py
# https: // leetcode.com/problems/construct-smallest-number-from-di-string/description/?envType=daily-question&envId=2025-02-18
# Solution taken from: https://www.youtube.com/watch?v=GgN8d22BEf0&ab_channel=NeetCodeIO
# Did I understand properly? No


class Solution:
    def smallestNumber(self, pattern: str) -> str:
        res,stack= [],[]
        for i in range(len(pattern)+1):
            stack.append(i+1)

            while stack and (i==len(pattern) or pattern[i]=="I"):
                res.append(str(stack.pop()))

        return  "".join(res)
            
        

